### PR TITLE
8318603: Parallelize sun/java2d/marlin/ClipShapeTest.java

### PR DESCRIPTION
**Pull request by mrserb**
The ClipShapeTest test is split into 4 different tests which could be executed in parallel.

The execution time is changed from:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/sun/java2d/marlin/ClipShapeTest.java
                                                         1     1     0     0   
==============================
TEST SUCCESS

Finished building target 'run-test' in configuration 'macosx-aarch64-server-release'

real	2m58.673s
user	3m8.847s
sys	0m11.845s
```

to:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/sun/java2d/marlin/ClipShapeTest.java
                                                         4     4     0     0   
==============================
TEST SUCCESS

Finished building target 'run-test' in configuration 'macosx-aarch64-server-release'

real	1m17.752s
user	3m25.308s
sys	0m12.987s

```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318603](https://bugs.openjdk.org/browse/JDK-8318603): Parallelize sun/java2d/marlin/ClipShapeTest.java (**Enhancement** - P4)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [67e3d2c5](https://git.openjdk.org/jdk/pull/17719/files/67e3d2c5548c3a5f8e0bbec6ac60ed190770dc3c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17719/head:pull/17719` \
`$ git checkout pull/17719`

Update a local copy of the PR: \
`$ git checkout pull/17719` \
`$ git pull https://git.openjdk.org/jdk.git pull/17719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17719`

View PR using the GUI difftool: \
`$ git pr show -t 17719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17719.diff">https://git.openjdk.org/jdk/pull/17719.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17719#issuecomment-1928982003)

### CODE REVIEW

**Comment by bourgesl** [comment]
LGTM, not an official reviewer.

I can check this test to select the minimal needed cases to detect regression (132 combinations for now) and reduce even more test wallclock time.

File: test/jdk/sun/java2d/marlin/ClipShapeTest.java [hunk 1]
```diff
@@ -62,10 +62,43 @@
...
+
+/*
+ * @test id=PolyDoDash
+ * @bug 8191814
+ * @summary Verifies that Marlin rendering generates the same
+ * images with and without clipping optimization with all possible
+ * stroke (cap/join) and/or dashes or fill modes (EO rules)
+ * for paths made of either 9 lines, 4 quads, 2 cubics (random)
+ * Note: Use the argument -slow to run more intensive tests (too much time)
+ *
```
Comment (shipilev) [inline test/jdk/sun/java2d/marlin/ClipShapeTest.java:75]: I think our style is only put these in the first `@run` block, and the rest to only carry the essentials for the tests to run. Simplifies the patch as well.

Comment (aivanov-jdk) [reply test/jdk/sun/java2d/marlin/ClipShapeTest.java:75]: Especially with such a long description that spans four lines.

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me except for the minor comment.

Do you mind updating the copyright year?

**Comment by mrserb** [comment]
duplicated comments are reworked

**Comment by bourgesl** [comment]
LGTM (comments fixed)

**Review by shipilev — APPROVED** [review]
All right, this looks good, thanks!

File: test/jdk/sun/java2d/marlin/ClipShapeTest.java [hunk 2]
```diff
@@ -55,50 +55,38 @@
...
+ * @summary Runs the test with "-cubic -doDash" options
  * @run main/othervm/timeout=300 -Dsun.java2d.renderer=sun.java2d.marlin.DMarlinRenderingEngine ClipShapeTest -cubic -doDash
  */
+
+/**
+ * Verifies that Marlin rendering generates the same images with and without
+ * clipping optimization with all possible stroke (cap/join) and/or dashes or
+ * fill modes (EO rules) for paths made of either 9 lines, 4 quads, 2 cubics
+ * (random)
+ * Note: Use the argument -slow to run more intensive tests (too much time)
```
Comment (aivanov-jdk) [inline test/jdk/sun/java2d/marlin/ClipShapeTest.java:88]:
```suggestion
 * <p>
 * Note: Use the argument {@code -slow} to run more intensive tests (too much time)
```
Adding <p> will render the note in a new paragraph when viewed in IDE (or on class name hover).

**Review by aivanov-jdk — APPROVED** [review]

**Comment by mrserb** [comment]
> I ran the test in our CI and I didn't notice an improvement. On the other hand, having separate test ids is more flexible.

I have retested by the complete jdk_desktop group of tests:
Base: 
```
real    6m57.901s
user    123m45.204s
sys     9m16.735s
```
Fix:
```
real    4m23.600s
user    126m35.078s
sys     8m50.104s

```
So it might depend on the system and used concurrency.

**Comment by prrace** [comment]
> I ran the test in our CI and I didn't notice an improvement. 

Did you run jdk:tier4 like Sergey was doing ?

Our CI runs some number of (headless) tests in parallel (I am not sure but think it is by default 4) and this might change
the whole dynamics of "which test is run last", compared to other test frameworks.

I don't know if whatever Sergey is using does the same, but they could be quite different in the number of CPUs per VM.

If you ran just the *one* test and it was no faster (wall-clock time) that suggests the (sub-)tests were not being run in parallel which would be interesting .. and we might want to ask jtreg folks for input.

**Review by aivanov-jdk — APPROVED** [review]

**Comment by aivanov-jdk** [comment]
> > I ran the test in our CI and I didn't notice an improvement.
> 
> Did you run jdk:tier4 like Sergey was doing ?

No, I ran just this single test. I don't know how Sergey is running the test.

> Our CI runs some number of (headless) tests in parallel (I am not sure but think it is by default 4) and this might change the whole dynamics of "which test is run last", compared to other test frameworks.

Yep, I know that headless tests are run in parallel, which makes perfect sense.

> I don't know if whatever Sergey is using does the same, but they could be quite different in the number of CPUs per VM.
> 
> If you ran just the _one_ test and it was no faster (wall-clock time) that suggests the (sub-)tests were not being run in parallel which would be interesting .. and we might want to ask jtreg folks for input.

Yes, I submitted a job with one test only.

At the same time, I didn't look thoroughly into the logs.

This .java file contains four different tests which can be run in parallel, it could save the overall time; in the worst case, the same four tests are run consecutively. No negative impact, so I support the change.

**Pull request closed without merging**